src/QsCompiler/CompilerOptimizations/FunctionEvaluation.fs

```diff
@@ -0,0 +1,145 @@
+﻿module Microsoft.Quantum.QsCompiler.CompilerOptimization.FunctionEvaluation
+
+open System.Collections.Immutable
+open Microsoft.Quantum.QsCompiler.DataTypes
+open Microsoft.Quantum.QsCompiler.SyntaxTokens
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+
+open Utils
+
+
+/// The current state of a function evaluation
+type FunctionState =
+| Normal
+| Returned of Expr
+| Failed of Expr
+| CouldNotEvaluate of string
+
+/// The current state of an expression evaluation
+type ExpressionResult<'A> =
+| ExprValue of 'A
+| ExprError of string
+
+
+/// Evaluates functions by stepping through their code
+type [<AbstractClass>] FunctionEvaluator(cd: CallableDict) =
+
+    /// Transforms a BoolLiteral into the corresponding bool
+    let castToBool x =
+        match x with
+        | BoolLiteral b -> ExprValue b
+        | _ -> "Not a BoolLiteral: " + (prettyPrint x) |> ExprError
+
+    /// Returns whether a FunctionState is an interrupt
+    let isInterrupt = function Normal -> false | _ -> true
+
+    /// The callback for the subroutine that evaluates and simplifies an expression
+    abstract member evaluateExpression: VariablesDict -> TypedExpression -> Expr
+
+    /// Evaluates a single Q# statement
+    member this.evaluateStatement (vars: VariablesDict) (statement: QsStatement): FunctionState =
+        match statement.Statement with
+        | QsExpressionStatement expr ->
+            this.evaluateExpression vars expr |> ignore; Normal
+        | QsReturnStatement expr ->
+            this.evaluateExpression vars expr |> Returned
+        | QsFailStatement expr ->
+            this.evaluateExpression vars expr |> Failed
+        | QsVariableDeclaration s ->
+            fillVars vars (StringTuple.fromSymbolTuple s.Lhs, this.evaluateExpression vars s.Rhs); Normal
+        | QsValueUpdate s ->
+            match s.Lhs.Expression with
+                | Identifier (LocalVariable name, tArgs) -> vars.setVar(name.Value, this.evaluateExpression vars s.Rhs); Normal
+                | _ -> "Unknown LHS of value update statement: " + (prettyPrint s.Lhs.Expression) |> CouldNotEvaluate
+        | QsConditionalStatement s ->
+            match s.ConditionalBlocks |>
+                Seq.map (fun (ts, block) -> this.evaluateExpression vars ts |> castToBool, block) |>
+                Seq.tryFind (fst >> function ExprValue true -> true | ExprValue false -> false | ExprError _ -> true), s.Default with
+            | Some (ExprError s, _), _ -> CouldNotEvaluate s
+            | Some (ExprValue _, block), _ | None, Value block -> this.evaluateScope vars block.Body
```

Comment:
I think you're misreading the code. If the evaluator cannot determine whether a case is executed, it will give an ExprError, which the first match case will report as a CoundNotEvaluate. The (None, Value block) represents knowing with certainty that none of the cases are evaluated, so we evaluate the default block instead.